src/OpenTelemetry/Batch.cs

```diff
@@ -140,6 +145,29 @@ public struct Enumerator : IEnumerator<T>
                 return false;
             };
 
+            private static readonly BatchEnumeratorMoveNextFunc MoveNextCircularBufferLogRecord = (ref Enumerator enumerator) =>
+            {
+                var circularBuffer = enumerator.circularBuffer;
+
+                var currentItem = enumerator.Current;
+                if (currentItem != null)
+                {
+                    if (typeof(T) == typeof(LogRecord))
```

Comment:
Is this check required here?

The Enumerator ctor already has this check:
```csharp
this.moveNextFunc = typeof(T) == typeof(LogRecord) ? MoveNextCircularBufferLogRecord : MoveNextCircularBuffer;
```

Reply:
@utpilla I did a little more investigation on this.

Ran this through SharpLab:

```csharp
[JitGeneric(typeof(LogRecord))]
[JitGeneric(typeof(Activity))]
public class Helpers<T>
{
   public void DisposeActionLogRecord(T item)
   {
      ((LogRecord)(object)item).Cleanup();
   }  
   
   public void DisposeActionLogRecordWithCheck(T item)
   {
      if (typeof(T) == typeof(LogRecord))
      {
          ((LogRecord)(object)item).Cleanup();
      }
   }     
    
   public void DisposeAction(T item)
   {
   }
}
```

For T == LogRecord...

```
Helpers`1[[LogRecord, _]].DisposeActionLogRecord(LogRecord)
    L0000: sub rsp, 0x28
    L0004: xor eax, eax
    L0006: mov [rsp+0x20], rax
    L000b: mov [rsp+0x20], dl
    L000f: lea rcx, [rsp+0x20]
    L0014: call 0x00007ffd197804d0
    L0019: nop
    L001a: add rsp, 0x28
    L001e: ret

Helpers`1[[LogRecord, _]].DisposeActionLogRecordWithCheck(LogRecord)
    L0000: sub rsp, 0x28
    L0004: xor eax, eax
    L0006: mov [rsp+0x20], rax
    L000b: mov [rsp+0x20], dl
    L000f: lea rcx, [rsp+0x20]
    L0014: call 0x00007ffd197804d0
    L0019: nop
    L001a: add rsp, 0x28
    L001e: ret

Helpers`1[[LogRecord, _]].DisposeAction(LogRecord)
    L0000: ret
```

Method with and without the type check look the same.

For T == Activity...

```
Helpers`1[[Activity, _]].DisposeActionLogRecord(Activity)
    L0000: push rdi
    L0001: push rsi
    L0002: sub rsp, 0x28
    L0006: xor eax, eax
    L0008: mov [rsp+0x20], rax
    L000d: mov esi, edx
    L000f: mov rcx, 0x7ffd1978d278
    L0019: call 0x00007ffd6e75b080
    L001e: mov rdi, rax
    L0021: mov [rdi+8], sil
    L0025: mov rdx, rdi
    L0028: mov rcx, 0x7ffd1978d178
    L0032: call 0x00007ffd0ebb7478
    L0037: movzx ecx, byte ptr [rdi+8]
    L003b: mov [rsp+0x20], cl
    L003f: lea rcx, [rsp+0x20]
    L0044: call 0x00007ffd197804d0
    L0049: nop
    L004a: add rsp, 0x28
    L004e: pop rsi
    L004f: pop rdi
    L0050: ret

Helpers`1[[Activity, _]].DisposeActionLogRecordWithCheck(Activity)
    L0000: ret

Helpers`1[[Activity, _]].DisposeAction(Activity)
    L0000: ret
```

Interesting, eh? Without the type check the JIT generated a lot of stuff. Probably a cast with a throw if the cast fails.

I'm thinking I'm going to change the code from...

```csharp
            private static readonly BatchEnumeratorMoveNextFunc MoveNextCircularBufferLogRecord = (ref Enumerator enumerator) =>
            {
                var circularBuffer = enumerator.circularBuffer;

                var currentItem = enumerator.Current;
                if (currentItem != null)
                {
                    if (typeof(T) == typeof(LogRecord))
                    {
                        LogRecordSharedPool.Current.Return((LogRecord)(object)currentItem);
                    }
                }

                if (circularBuffer!.RemovedCount < enumerator.targetCount)
                {
                    enumerator.current = circularBuffer.Read();
                    return true;
                }

                enumerator.current = null;
                return false;
            };
```

...to...

```csharp
            private static readonly BatchEnumeratorMoveNextFunc MoveNextCircularBufferLogRecord = (ref Enumerator enumerator) =>
            {
                if (typeof(T) == typeof(LogRecord))
                {
                    var circularBuffer = enumerator.circularBuffer;

                    var currentItem = enumerator.Current;
                    if (currentItem != null)
                    {
                        LogRecordSharedPool.Current.Return((LogRecord)(object)currentItem);
                    }

                    if (circularBuffer!.RemovedCount < enumerator.targetCount)
                    {
                        enumerator.current = circularBuffer.Read();
                        return true;
                    }

                    enumerator.current = null;
                }

                return false;
            };
```

By placing the type check over the whole block the JIT _should_ understand it doesn't need to generate anything for batches of activity/metric which will lead to less memory being taken up by code that will never execute. How you feel about that?

Reply:
Great finding!